// @gorhom/bottom-sheet/lib/commonjs/hooks/useGestureHandler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useGestureHandler = void 0;

var _reactNativeReanimated = require("react-native-reanimated");

const useGestureHandler = (type, state, handleOnStart, handleOnActive, handleOnEnd) => {
  const gestureHandler = (0, _reactNativeReanimated.useAnimatedGestureHandler)({
    onStart: (payload, context) => {
      state.value = payload.state;
      handleOnStart(type, payload, context);
    },
    onActive: (payload, context) => {
      state.value = payload.state;
      handleOnActive(type, payload, context);
    },
    onEnd: (payload, context) => {
      state.value = payload.state;
      handleOnEnd(type, payload, context);
    },
    onCancel: payload => {
      state.value = payload.state;
    },
    onFail: payload => {
      state.value = payload.state;
    },
    onFinish: payload => {
      state.value = payload.state;
    }
  }, [type, state, handleOnStart, handleOnActive, handleOnEnd]);
  return gestureHandler;
};

exports.useGestureHandler = useGestureHandler;
